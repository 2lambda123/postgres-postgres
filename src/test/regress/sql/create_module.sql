CREATE TABLE cm_test (a int, b text);

CREATE MODULE mtest1
  CREATE FUNCTION m1testa() RETURNS text
     LANGUAGE sql
     RETURN '1x'
  CREATE FUNCTION m1testb() RETURNS text
     LANGUAGE sql
     RETURN '1y';

CREATE SCHEMA temp_mod_test;
GRANT ALL ON SCHEMA temp_mod_test TO public;

CREATE MODULE temp_mod_test.mtest2
  CREATE PROCEDURE m2testa(x text)
  LANGUAGE SQL
  AS $$
  INSERT INTO cm_test VALUES (1, x);
  $$
  CREATE FUNCTION m2testb() RETURNS text
     LANGUAGE sql
     RETURN '2y';

CREATE MODULE mtest3
  CREATE FUNCTION mtest3.m3testa() RETURNS text
     LANGUAGE sql
     RETURN '3x';

SELECT mtest1.m1testa();
SELECT mtest1.m1testb();

SELECT public.mtest1.m1testa();
SELECT public.mtest1.m1testb();

SELECT temp_mod_test.mtest2.m2testb();

SELECT temp_mod_test.mtest2.m2testa('x');  -- error
CALL temp_mod_test.mtest2.m2testa('a');  -- ok
CALL temp_mod_test.mtest2.m2testa('xy' || 'zzy');  -- ok, constant-folded arg

CREATE PROCEDURE mtest1.m1testc(x text)
  LANGUAGE SQL
  AS $$
  INSERT INTO cm_test VALUES (2, x);
  $$;

CALL mtest1.m1testc('a');  -- ok

DROP PROCEDURE mtest1.m1testc(text);
DROP FUNCTION temp_mod_test.mtest2.m2testb();

-- cleanup

DROP MODULE mtest1 CASCADE;
DROP MODULE temp_mod_test.mtest2 CASCADE;

DROP SCHEMA temp_mod_test;
DROP TABLE cm_test;
