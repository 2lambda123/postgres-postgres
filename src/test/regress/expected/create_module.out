CREATE TABLE cm_test (a int, b text);
CREATE MODULE mtest1
  CREATE FUNCTION m1testa() RETURNS text
     LANGUAGE sql
     RETURN '1x'
  CREATE FUNCTION m1testb() RETURNS text
     LANGUAGE sql
     RETURN '1y';
CREATE SCHEMA temp_mod_test;
GRANT ALL ON SCHEMA temp_mod_test TO public;
CREATE MODULE temp_mod_test.mtest2
  CREATE PROCEDURE m2testa(x text)
  LANGUAGE SQL
  AS $$
  INSERT INTO cm_test VALUES (1, x);
  $$
  CREATE FUNCTION m2testb() RETURNS text
     LANGUAGE sql
     RETURN '2y';
CREATE MODULE mtest3
  CREATE FUNCTION mtest3.m3testa() RETURNS text
     LANGUAGE sql
     RETURN '3x';
ERROR:  CREATE FUNCTION (mtest3.m3testa) specifies a namespace inside of CREATE MODULE (mtest3)
SELECT mtest1.m1testa();
 m1testa 
---------
 1x
(1 row)

SELECT mtest1.m1testb();
 m1testb 
---------
 1y
(1 row)

SELECT public.mtest1.m1testa();
 m1testa 
---------
 1x
(1 row)

SELECT public.mtest1.m1testb();
 m1testb 
---------
 1y
(1 row)

SELECT temp_mod_test.mtest2.m2testb();
 m2testb 
---------
 2y
(1 row)

SELECT temp_mod_test.mtest2.m2testa('x');  -- error
ERROR:  temp_mod_test.mtest2.m2testa(unknown) is a procedure
LINE 1: SELECT temp_mod_test.mtest2.m2testa('x');
               ^
HINT:  To call a procedure, use CALL.
CALL temp_mod_test.mtest2.m2testa('a');  -- ok
CALL temp_mod_test.mtest2.m2testa('xy' || 'zzy');  -- ok, constant-folded arg
CREATE PROCEDURE mtest1.m1testc(x text)
  LANGUAGE SQL
  AS $$
  INSERT INTO cm_test VALUES (2, x);
  $$;
ERROR:  schema "mtest1" does not exist
CALL mtest1.m1testc('a');  -- ok
ERROR:  schema "mtest1" does not exist
LINE 1: CALL mtest1.m1testc('a');
             ^
DROP PROCEDURE mtest1.m1testc(text);
ERROR:  procedure mtest1.m1testc(text) does not exist
DROP FUNCTION temp_mod_test.mtest2.m2testb();
-- cleanup
DROP MODULE mtest1 CASCADE;
NOTICE:  drop cascades to 2 other objects
DETAIL:  drop cascades to function mtest1.m1testa()
drop cascades to function mtest1.m1testb()
DROP MODULE temp_mod_test.mtest2 CASCADE;
NOTICE:  drop cascades to function mtest2.m2testa(text)
DROP SCHEMA temp_mod_test;
DROP TABLE cm_test;
